### PR TITLE
impl(pubsub): new `AckHandler` to support exactly-once delivery

### PR DESCRIPTION
**Pull request by coryan**
This introduces the class in the `pubsub_internal` namespace. I will
weave this class through the implementation, and then move it to the
public namespace(s).

Part of the work for #9327

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9394)

### CODE REVIEW

File: google/cloud/pubsub/internal/exactly_once_ack_handler.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  ExactlyOnceAckHandler& operator=(ExactlyOnceAckHandler&&) = default;
+
+  /**
+   * Acknowledges the message associated with this handler.
+   *
+   * @par Idempotency
+   * Note that this is not an idempotent operation, and therefore it is never
+   * retried. Furthermore, the service may still resend a message after a
+   * successful `ack()`.
+   * Applications developers are reminded that Cloud Pub/Sub
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/exactly_once_ack_handler.h:63]: nit: weird line break

**Review by dbolduc — APPROVED** [review]

File: google/cloud/pubsub/internal/exactly_once_ack_handler.h [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * Defines the interface to acknowledge and reject messages.
+ *
+ * When applications register a callback to receive Pub/Sub messages the
+ * callback must be able to receive both a `pubsub::Message` and its associated
+ * `pubsub::AckHandler`. Actions on a `pubsub::AckHandler` always affect the
+ * same message received in the callback. Applications cannot create standalone
+ * handlers (except in unit tests via mocks).
+ *
+ * This interface allows applications to acknowledge and reject messages in a
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/exactly_once_ack_handler.h]: Fixed.

File: google/cloud/pubsub/internal/exactly_once_ack_handler.h (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/internal/exactly_once_ack_handler.h]: That made me re-read the comments and they were wrong for this class.  Can you take another look?

**Review by dbolduc — APPROVED** [review]

File: google/cloud/pubsub/internal/exactly_once_ack_handler.h (hunk 1 above)
Comment (dbolduc) [reply google/cloud/pubsub/internal/exactly_once_ack_handler.h]: LGTM

File: google/cloud/pubsub/ack_handler.h [hunk 3]
```diff
@@ -33,7 +33,7 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
  * same message received in the callback. Applications cannot create standalone
  * handlers (except in unit tests via mocks).
  *
- * This interface allows applications to acknowledge and reject messages in a
+ * This interface allows applications to acknowledge and reject messages that a
```
Comment (devbww) [inline google/cloud/pubsub/ack_handler.h:36]: `s/that a/that are/`?

**Pull request merged**